aptoide-installer/src/main/java/cm/aptoide/pt/installer/fetch/FetchDownloader.kt

```
@@ -0,0 +1,215 @@
+      }
+      installPermissions.checkIfCanInstall()
+      emit(Unit)
+    }.flatMapLatest {
+      downloadFiles(packageName, versionCode, installationFiles)
+    }.map {
+      coroutineContext.ensureActive()
+      (it * 100).toInt()
+    }
+  }
+    .also { InstallerWorker.enqueue(context) }
+
+  private fun downloadFiles(
+    packageName: String,
+    versionCode: Long,
+    installationFiles: Set<InstallationFile>
+  ): Flow<Double> {
+    val totalSize = installationFiles.sumOf { it.fileSize }.toDouble()
+    val fetch = Fetch.Impl.getDefaultInstance()
+
+    val destinationDir = File(downloadsPath, packageName).apply {
+      if (!exists()) {
+        mkdirs().let {
+          if (!it) throw IllegalStateException("Can't create download folder: $downloadsPath/$packageName")
+        }
+      }
+    }
+
+    val requests = installationFiles.mapNotNull { file ->
+      File(destinationDir, file.name)
+        .takeUnless { it.checkMd5(file.md5) }
+        ?.let {
+          Request(file.url, it.toUri())
+            .apply {
+              priority = Priority.NORMAL
+              networkType = NetworkType.ALL
+              groupId = packageName.asGroupId()
+              addHeader(VERSION_CODE, versionCode.toString())
+              addHeader(PACKAGE, packageName)
+              addHeader(FILE_TYPE, file.type.toString())
+            }
+            .also {
+              contentHashes.put(it.id, file.md5)
+            }
+        }
+    }
+
+    val downloads = requests.map { it.id }.toMutableSet()
+    val downloadsProgresses = requests.associate { it.id to 0L }.toMutableMap()
+
+    return callbackFlow<Double> {
+      try {
+        trySend(0.0)
+
+        if (downloads.isEmpty()) {
+          trySend(1.0)
+          close()
+        }
+
+        fetch.enqueue(requests) { enqueuedRequests ->
+          enqueuedRequests.find { it.second != Error.NONE }?.let {
+            close(it.second.throwable ?: IllegalStateException("Error enqueuing requests"))
```

should we close if there is an error ?
How do they know if there are "repeated files", what is the id ?
Couldnt we ignore the repeated one instead of giving error ?

---

- If there is an error enqueuing one of the requests, then that file will not be downloaded so it is probably better to throw an error asap.
- Not 100% sure, but if the request local file uri is the same then Fetch will treat it as a duplicate/repeated file.
- The request id is generated automatically by Fetch.
- There should be no repeated ones, since each request is created based on the installation files

---

i thought the error while enqueuing was always associated with duplication. So that is why i was saying we could just ignore